streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.scala
+package kstream
+
+import org.apache.kafka.streams.kstream.{KGroupedStream => KGroupedStreamJ, _}
+import org.apache.kafka.common.serialization.Serde
+import org.apache.kafka.streams.scala.ImplicitConversions._
+import org.apache.kafka.streams.scala.FunctionConversions._
+
+
+/**
+ * Wraps the Java class KGroupedStream and delegates method calls to the underlying Java object.
+ */
+class KGroupedStream[K, V](inner: KGroupedStreamJ[K, V]) {
+
+  def count(): KTable[K, Long] = {
+    val c: KTable[K, java.lang.Long] = inner.count()
+    c.mapValues[Long](Long2long _)
+  }
+
+  def count(store: String, keySerde: Option[Serde[K]] = None): KTable[K, Long] = {
```

Comment:
Why do we only allow to specify a `keySerde` but not replace the store with a different one?

Scala noob question: would it be possible to have a single `count` / `reduce` etc instead of overloads and use `Option` and implicits to infer optional arguments?

Reply:
@mjsax -

I think I may be missing something here. We are allowing the user to pass in the `store` *and* `keySerde`. And we create the `Materialized` out of the 2 with `Long` as the value serde.

However we were allowing the user to pass in the `keySerde` optionally and in case the user does not supply one we assumed it will be taken from the config. This is actually a remnant from the earlier thoughts where we thought passing serdes through config may be a good idea. However in the current context, we should not make `keySerde` optional. Here's the suggestion for the changed API ..

```scala
def count(store: String, keySerde: Serde[K]): KTable[K, Long] = { 
  val materialized = Materialized.as[K, java.lang.Long, ByteArrayKeyValueStore](store).withKeySerde(keySerde)
  val c: KTable[K, java.lang.Long] = inner.count(materialized)
  c.mapValues[Long](Long2long _)
}
```

An alternative option could have been to allow the user to pass in the `Materialized` instance itself (like we do in the `reduce` function). The problem with that alternative is that the Java API expects `java.lang.Long` as the value serde, while the Scala API needs to take a `scala.Long`. And there is no way we can convert a `Materialized.as[K, scala.Long, ByteArrayKeyValueStore]` to `Materialized.as[K, java.lang.Long, ByteArrayKeyValueStore]` without a cast. Hence the divergence in the API signature between `count` and `reduce`.

Please share if u have any other thoughts.

Reply:
That makes sense.

My 2 cents:
We're presenting the scala KGroupedStream basically *as* the java one, but not implementing the interface so that we can smooth over a couple of specific gaps. I think this is a good call, but it's also a huge risk for cognitive dissonance and developer confusion, since they will read the java version of the docs and try to use that knowledge in scala. Therefore, it's important to be super disciplined about making sure the methods available are as close to the java interface as possible.

Clearly, moving serdes, etc., to implicit params is the kind of thing we *do* want to do. But I think that presenting `count(String,Serde[K])` instead of `count(Materialized[K, Long, KeyValueStore[Bytes, Array[Byte]]]` is too far off.

I do agree that the method should take a scala type. Apparently, it's perfectly fine to cast `Serde[scala.Long]` to `Serde[java.Long]`. Does that same logic apply here? Alternatively, we can actually convert the `Materialized[java]` to a `Materialized[scala]`.

Reply:
@vvcephei .. Sure we can do the following instead ..

```scala
def count(materialized: Materialized[K, Long, ByteArrayKeyValueStore]): KTable[K, Long] = { 
  val c: KTable[K, java.lang.Long] = inner.count(materialized.asInstanceOf[Materialized[K, java.lang.Long, ByteArrayKeyValueStore]])
  c.mapValues[Long](Long2long _)
}
```

WDYT ?

Reply:
Ok, I'm already a little nervous about the cast in one direction, so this feels super gross, but would this work?
```scala
def count(materialized: Materialized[K, Long, ByteArrayKeyValueStore]): KTable[K, Long] = { 
  inner.count(materialized.asInstanceOf[Materialized[K, java.lang.Long, ByteArrayKeyValueStore]]).asInstanceOf[KTable[K, scala.Long]
}
```
Please understand I'm wincing as I type this.

Reply:
Well, we allowing to pass in a store "name" (String) but not a store. Note, that `Materialized` allows to replace default RocksDB with an in-memory story, disable change-capture-logging or even use a custom store implementation.

Reply:
How about the API that I suggested above ? It takes Materialized much like the Java API though we need a cast. 

@vvcephei - in my implementation we have 1 cast and the other map for the long conversion in KTable.

Reply:
@debasishg Sorry, I should have acked your implementation. I was actually proposing an evolution of it.

It just seems a bit unfortunate to have to add a real function invocation to the topology in order to do the cast back to `scala.Long`. The version I proposed just does a cast back out without adding anything new to the topology. Does that make sense?

At the risk of sounding like an idiot, if it's fine to do the cast on the way in, then it should be fine again on the way out, right?

Reply:
@vvcephei - cast is a runtime operation and my philosophy is to minimize its use. And `scala.Predef` indeed uses `Long2long` to do such conversions. Hence I would like to prefer using proper functions when available instead of the cast.